# Importing the needed packages

In [2]:
import pandas as pd
pd.options.mode.chained_assignment = None  # default='warn'

# Data Access and Profiling:
###     Importing the Passengers Satisfaction Survey dataset

In [13]:
df= pd.read_csv ('Passengers_Satisfaction_Survey.csv', low_memory=False) 
df

Gender      Customer Type   Age   Type of Travel     Class  \
0         Male     Loyal Customer  69.0  Personal Travel  Business   
1         Male  Disloyal Customer  20.0  Business travel       Eco   
2         Male  Disloyal Customer  41.0  Business travel  Business   
3         Male     Loyal Customer  52.0  Business travel  Business   
4       Female  Disloyal Customer  21.0  Business travel  Business   
...        ...                ...   ...              ...       ...   
104905    Male     Loyal Customer  45.0  Business travel  Business   
104906  Female     Loyal Customer  35.0  Business travel  Business   
104907    Male     Loyal Customer  47.0  Business travel  Business   
104908    Male     Loyal Customer  42.0  Business travel  Business   
104909  Female           disloyal   0.0  Business travel  Business   

        Flight Distance  Inflight wifi service  \
0                  32.0                    3.0   
1                   0.0                    3.0   
2                   0.0                    1.0   
3                   0.0                    2.0   
4                   0.0                    4.0   
...                 ...                    ...   
104905           3305.0                    2.0   
104906            337.0                    5.0   
104907           3496.0                    2.0   
104908           3605.0                    4.0   
104909            862.0                    5.0   

        Departure/Arrival time convenient Ease of Online booking  \
0                                     2.0                      3   
1                                     3.0                      3   
2                                     1.0                    NaN   
3                                     4.0                      4   
4                                     0.0                    NaN   
...                                   ...                    ...   
104905                                2.0                      2   
104906                                5.0                      5   
104907                                2.0                      2   
104908                                4.0                      4   
104909                                5.0                      5   

       Food and drink  Inflight entertainment Baggage handling  Cleanliness  \
0                   5                     1.0                1          3.0   
1                   2                    10.0                4          2.0   
2                 NaN                     NaN              NaN          NaN   
3                   1                     2.0                2          4.0   
4                 NaN                     NaN              NaN          NaN   
...               ...                     ...              ...          ...   
104905              3                     4.0                4          3.0   
104906              2                     4.0                4          5.0   
104907              2                     5.0                5          4.0   
104908              3                     4.0                4          3.0   
104909              5                     5.0                1          5.0   

        Departure Delay in Minutes Arrival Delay in Minutes  Satisfaction  
0                             14.0                        0  Dissatisfied  
1                              0.0                        0  dissatisfied  
2                              NaN                      NaN       neutral  
3                              0.0                        6  Dissatisfied  
4                              NaN                      NaN       neutral  
...                            ...                      ...           ...  
104905                         8.0                        0     satisfied  
104906                        54.0                       40     satisfied  
104907                         0.0                        2     satisfied  
104908                         0.0      

In [17]:
df.shape
RowsCount = df.shape[0]
ColumnsCount = df.shape[1] 

In [19]:
print(f"The dataset contains {RowsCount} rows and {ColumnsCount} columns.")

The dataset contains 104910 rows and 16 columns.


###     Creating subset Dataframe of Business Class

In [350]:
# Creating new dataframe
df_Business = df[(df["Class"] == "Business")] 
df_Business.head()

Gender      Customer Type   Age   Type of Travel     Class  \
0    Male     Loyal Customer  69.0  Personal Travel  Business   
2    Male  Disloyal Customer  41.0  Business travel  Business   
3    Male     Loyal Customer  52.0  Business travel  Business   
4  Female  Disloyal Customer  21.0  Business travel  Business   
5  Female     Loyal Customer  57.0  Business travel  Business   

   Flight Distance  Inflight wifi service  Departure/Arrival time convenient  \
0             32.0                    3.0                                2.0   
2              0.0                    1.0                                1.0   
3              0.0                    2.0                                4.0   
4              0.0                    4.0                                0.0   
5              0.0                    3.0                                4.0   

  Ease of Online booking Food and drink  Inflight entertainment  \
0                      3              5                     1.0   
2                    NaN            NaN                     NaN   
3                      4              1                     2.0   
4                    NaN            NaN                     NaN   
5                      4              5                     2.0   

  Baggage handling  Cleanliness  Departure Delay in Minutes  \
0                1          3.0                        14.0   
2              NaN          NaN                         NaN   
3                2          4.0                         0.0   
4              NaN          NaN                         NaN   
5                3          2.0                         0.0   

  Arrival Delay in Minutes  Satisfaction  
0                        0  Dissatisfied  
2                      NaN       neutral  
3                        6  Dissatisfied  
4                      NaN       neutral  
5                        0  Dissatisfied

In [38]:
df_Business.shape
df_Business_Rows = df_Business.shape[0]
df_Business_Columns = df_Business.shape[1]

In [42]:
print(f"Business Class dataset contains {df_Business_Rows} rows and {df_Business_Columns} columns.")

Business Class dataset contains 50000 rows and 16 columns.


In [44]:
df_Business.info()

<class 'pandas.core.frame.DataFrame'>
Index: 50000 entries, 0 to 104909
Data columns (total 16 columns):
 #   Column                             Non-Null Count  Dtype  
---  ------                             --------------  -----  
 0   Gender                             48239 non-null  object 
 1   Customer Type                      48239 non-null  object 
 2   Age                                47373 non-null  float64
 3   Type of Travel                     48239 non-null  object 
 4   Class                              50000 non-null  object 
 5   Flight Distance                    46604 non-null  float64
 6   Inflight wifi service              48239 non-null  float64
 7   Departure/Arrival time convenient  48239 non-null  float64
 8   Ease of Online booking             42705 non-null  object 
 9   Food and drink                     42705 non-null  object 
 10  Inflight entertainment             43649 non-null  float64
 11  Baggage handling                   43649 non-null  object 

### Checking number of unique values and unique values in each column

In [52]:
cols = df_Business.columns
for col in cols:
    print(col)
    unique = df_Business[col].unique()
    print('Number of unique values:',len(unique))
    if(len(unique) < 20):
        print(unique, '\n====================================\n\n')

Gender
Number of unique values: 6
['Male' 'Female' 'F' 'M' nan 'Fe male'] 


Customer Type
Number of unique values: 5
['Loyal Customer' 'Disloyal Customer' nan 'disloyal' 'Loyal'] 


Age
Number of unique values: 75
Type of Travel
Number of unique values: 4
['Personal Travel' 'Business travel' 'Business' nan] 


Class
Number of unique values: 1
['Business'] 


Flight Distance
Number of unique values: 3575
Inflight wifi service
Number of unique values: 7
[ 3.  1.  2.  4.  0.  5. nan] 


Departure/Arrival time convenient
Number of unique values: 7
[ 2.  1.  4.  0.  5.  3. nan] 


Ease of Online booking
Number of unique values: 8
['3' nan '4' '0' '5' 'Unknown' '2' '1'] 


Food and drink
Number of unique values: 8
['5' nan '1' '3' 'Great' '4' '2' '0'] 


Inflight entertainment
Number of unique values: 8
[ 1. nan  2.  3.  4.  5. 10.  0.] 


Baggage handling
Number of unique values: 7
['1' nan '2' '3' '5' '4' 'A'] 


Cleanliness
Number of unique values: 6
[ 3. nan  4.  2.  5.  0.] 


Departur

In [229]:
df_Business.isnull().sum()

Gender                               1761
Customer Type                        1761
Age                                  2627
Type of Travel                       1761
Class                                   0
Flight Distance                      3396
Inflight wifi service                1761
Departure/Arrival time convenient    1761
Ease of Online booking               7295
Food and drink                       7295
Inflight entertainment               6351
Baggage handling                     6351
Cleanliness                          6351
Departure Delay in Minutes           6351
Arrival Delay in Minutes             7902
Satisfaction                         2160
dtype: int64

#### There are missing values in 15 columns

## Summary statistics for each data column.

In [61]:
df_Business.describe(include='all') # for numeric columns

Gender   Customer Type           Age   Type of Travel     Class  \
count    48239           48239  47373.000000            48239     50000   
unique       5               4           NaN                3         1   
top     Female  Loyal Customer           NaN  Business travel  Business   
freq     23000           40770           NaN            44063     50000   
mean       NaN             NaN     41.853186              NaN       NaN   
std        NaN             NaN     13.730683              NaN       NaN   
min        NaN             NaN      0.000000              NaN       NaN   
25%        NaN             NaN     34.000000              NaN       NaN   
50%        NaN             NaN     43.000000              NaN       NaN   
75%        NaN             NaN     52.000000              NaN       NaN   
max        NaN             NaN    185.000000              NaN       NaN   

        Flight Distance  Inflight wifi service  \
count      4.660400e+04           48239.000000   
unique              NaN                    NaN   
top                 NaN                    NaN   
freq                NaN                    NaN   
mean       1.934729e+03               2.774519   
std        4.369344e+04               1.423917   
min       -5.421000e+03               0.000000   
25%        5.940000e+02               2.000000   
50%        1.597000e+03               3.000000   
75%        2.565000e+03               4.000000   
max        9.000000e+06               5.000000   

        Departure/Arrival time convenient Ease of Online booking  \
count                        48239.000000                  42705   
unique                                NaN                      7   
top                                   NaN                      3   
freq                                  NaN                   8635   
mean                             2.910695                    NaN   
std                              1.501711                    NaN   
min                              0.000000                    NaN   
25%                              2.000000                    NaN   
50%                              3.000000                    NaN   
75%                              4.000000                    NaN   
max                              5.000000                    NaN   

       Food and drink  Inflight entertainment Baggage handling   Cleanliness  \
count           42705            43649.000000            43649  43649.000000   
unique              7                     NaN                6           NaN   
top                 4                     NaN                4           NaN   
freq            10890                     NaN            17056           NaN   
mean              NaN                3.835391              NaN      3.737520   
std               NaN                1.233508              NaN      0.987664   
min               NaN                0.000000              NaN      0.000000   
25%               NaN                3.000000              NaN      3.000000   
50%               NaN                4.000000              NaN      4.000000   
75%               NaN                5.000000              NaN      5.000000   
max               NaN               10.000000              NaN      5.000000   

        Departure Delay in Minutes Arrival Delay in Minutes Satisfaction  
count                 43649.000000                    42098        47840  
unique                         NaN                      363            6  
top                            NaN                        0    satisfied  
freq                           NaN                    24229        30398  
mean                     13.595363                      NaN          NaN  
std                      36.595048                      NaN          NaN  
min                      -2.000000                      NaN          NaN  
25%                       0.000000                      NaN          NaN  
50%                       0.000000                   

## Frequency distribution for each column

In [64]:
cols = df_Business.columns
for col in cols:
    print(col)
    value_counts = df_Business[col].value_counts()
    print(value_counts, '\n\n=======================================================')

Gender
Gender
Female     23000
Male       22830
F           1361
M           1047
Fe male        1
Name: count, dtype: int64 

Customer Type
Customer Type
Loyal Customer       40770
Disloyal Customer     6636
disloyal               831
Loyal                    2
Name: count, dtype: int64 

Age
Age
39.0     1935
40.0     1677
41.0     1533
42.0     1514
44.0     1506
         ... 
73.0       22
78.0       17
85.0       14
180.0       1
185.0       1
Name: count, Length: 74, dtype: int64 

Type of Travel
Type of Travel
Business travel    44063
Business            2122
Personal Travel     2054
Name: count, dtype: int64 

Class
Class
Business    50000
Name: count, dtype: int64 

Flight Distance
Flight Distance
337.0        225
2475.0       187
2586.0       148
1744.0       133
404.0        133
            ... 
1285.0         1
1287.0         1
1288.0         1
1289.0         1
9000000.0      1
Name: count, Length: 3574, dtype: int64 

Inflight wifi service
Inflight wifi service
2.0    1078

### Checking for duplicate rows

In [71]:
df_Business.duplicated
duplicate = df_Business[df_Business.duplicated()]
  
print("Duplicate Rows :")
duplicate

Duplicate Rows :


Gender      Customer Type   Age   Type of Travel     Class  \
153          F  Disloyal Customer  25.0         Business  Business   
349        NaN                NaN   NaN              NaN  Business   
369        NaN                NaN   NaN              NaN  Business   
705        NaN                NaN   NaN              NaN  Business   
1159    Female     Loyal Customer  49.0  Business travel  Business   
...        ...                ...   ...              ...       ...   
104737  Female     Loyal Customer  37.0  Business travel  Business   
104768     NaN                NaN   NaN              NaN  Business   
104789    Male     Loyal Customer  24.0         Business  Business   
104825       M     Loyal Customer  30.0  Business travel  Business   
104840       M     Loyal Customer  30.0  Business travel  Business   

        Flight Distance  Inflight wifi service  \
153                67.0                    1.0   
349                 NaN                    NaN   
369                 NaN                    NaN   
705                 NaN                    NaN   
1159              101.0                    1.0   
...                 ...                    ...   
104737           3987.0                    3.0   
104768              NaN                    NaN   
104789           3993.0                    2.0   
104825              NaN                    3.0   
104840              NaN                    4.0   

        Departure/Arrival time convenient Ease of Online booking  \
153                                   0.0                      1   
349                                   NaN                    NaN   
369                                   NaN                    NaN   
705                                   NaN                    NaN   
1159                                  4.0                      4   
...                                   ...                    ...   
104737                                1.0                      3   
104768                                NaN                    NaN   
104789                                5.0                      5   
104825                                3.0                    NaN   
104840                                4.0                    NaN   

       Food and drink  Inflight entertainment Baggage handling  Cleanliness  \
153                 3                     3.0                4          3.0   
349               NaN                     NaN              NaN          NaN   
369               NaN                     NaN              NaN          NaN   
705               NaN                     NaN              NaN          NaN   
1159                3                     2.0                1          3.0   
...               ...                     ...              ...          ...   
104737              5                     5.0                5          5.0   
104768            NaN                     NaN              NaN          NaN   
104789              2                     2.0                A          2.0   
104825            NaN                     4.0                1          4.0   
104840            NaN                     4.0                4          4.0   

        Departure Delay in Minutes Arrival Delay in Minutes  Satisfaction  
153                            0.0                        0           NaN  
349                            NaN                      NaN  dissatisfied  
369                            NaN                      NaN  dissatisfied  
705                            NaN                      NaN  dissatisfied  
1159                          11.0                       14  Dissatisfied  
...                            ...                      ...           ...  
104737                        10.0                       23     satisfied  
104768                         NaN                      NaN  dissatisfied  
104789                        23.0                      -45  dissatisfied  
104825                        10.0      

#### There are 3064 duplicate entries in our dataset

---

## Data Profiling Summary 
### List of the issues that the data contains

Dataset Overview:
1. Initial dataset contained 104,910 rows and 16 columns.
2. A refined subset, df_Business, was created with 50,000 rows and 16 columns to focus on specific business-related analyses.
3. The dataset includes float and categorical data types.

Data Quality Issues:
1. Missing Values: All columns contain missing values, which need imputation or handling.
2. Duplicate Rows: Identified 3,064 duplicate rows that require removal or consolidation.

Feature-Level Analysis:
1. Gender: Contains inconsistent labels for gender representation. Standardize values to "Male" and "Female" and ensure the data type is either object or category.
2. Customer Type: Inconsistent values for customer categorization. Normalize entries to "Loyal Customer" and "Disloyal Customer."
3. Age: Contains outlier values (e.g., 0, 180, 185) that need correction to reflect realistic ages.
4. Type of Travel: Duplicate entries for "Business Travel" should be consolidated into a single, consistent label.

Outlier Identification:
1. Flight Distance: Contains extreme outlier values (e.g., 2.8 million and 9 million), which should be omitted for accurate analysis.
2. Departure Delay in Minutes and Arrival Delay in Minutes: Extreme values need to be capped or filtered to eliminate unrealistic data points.

Scale Inconsistencies:
1. Rating Columns: Multiple columns (e.g., Inflight Wifi Service, Departure/Arrival Time Convenience, Ease of Online Booking, Food and Drink, Inflight Entertainment, Baggage Handling, and Cleanliness) contain ratings outside the intended 1–5 scale. Values need correction to fall within this defined range.
2. Inflight Wifi Service: "0" ratings should be removed or corrected, with all ratings between 1 and 5.

Additional Data Consistency Issues:
1. Satisfaction: Duplicate or inconsistent values observed in this column. Standardization of satisfaction responses is required.



---
---
---






# Data  Cleansing

### Dropping Class Column from the df_Business dataset.

In [352]:
df_Business.drop(['Class'], axis=1, inplace=True)
df_Business = df_Business.reset_index(drop=True)  # To reset the df_Business index
df_Business

Gender      Customer Type   Age   Type of Travel  Flight Distance  \
0        Male     Loyal Customer  69.0  Personal Travel             32.0   
1        Male  Disloyal Customer  41.0  Business travel              0.0   
2        Male     Loyal Customer  52.0  Business travel              0.0   
3      Female  Disloyal Customer  21.0  Business travel              0.0   
4      Female     Loyal Customer  57.0  Business travel              0.0   
...       ...                ...   ...              ...              ...   
49995    Male     Loyal Customer  45.0  Business travel           3305.0   
49996  Female     Loyal Customer  35.0  Business travel            337.0   
49997    Male     Loyal Customer  47.0  Business travel           3496.0   
49998    Male     Loyal Customer  42.0  Business travel           3605.0   
49999  Female           disloyal   0.0  Business travel            862.0   

       Inflight wifi service  Departure/Arrival time convenient  \
0                        3.0                                2.0   
1                        1.0                                1.0   
2                        2.0                                4.0   
3                        4.0                                0.0   
4                        3.0                                4.0   
...                      ...                                ...   
49995                    2.0                                2.0   
49996                    5.0                                5.0   
49997                    2.0                                2.0   
49998                    4.0                                4.0   
49999                    5.0                                5.0   

      Ease of Online booking Food and drink  Inflight entertainment  \
0                          3              5                     1.0   
1                        NaN            NaN                     NaN   
2                          4              1                     2.0   
3                        NaN            NaN                     NaN   
4                          4              5                     2.0   
...                      ...            ...                     ...   
49995                      2              3                     4.0   
49996                      5              2                     4.0   
49997                      2              2                     5.0   
49998                      4              3                     4.0   
49999                      5              5                     5.0   

      Baggage handling  Cleanliness  Departure Delay in Minutes  \
0                    1          3.0                        14.0   
1                  NaN          NaN                         NaN   
2                    2          4.0                         0.0   
3                  NaN          NaN                         NaN   
4                    3          2.0                         0.0   
...                ...          ...                         ...   
49995                4          3.0                         8.0   
49996                4          5.0                        54.0   
49997                5          4.0                         0.0   
49998                4          3.0                         0.0   
49999                1          5.0                         9.0   

      Arrival Delay in Minutes  Satisfaction  
0                            0  Dissatisfied  
1                          NaN       neutral  
2                            6  Dissatisfied  
3                          NaN       neutral  
4                            0  Dissatisfied  
...                        ...           ...  
49995                        0     satisfied  
49996                       40     satisfied  
49997                        2     satisfied  
49998                        0     satisfied  
49999                        5       neutral  

[50000 rows x 15 columns]

#### List of Columns present in Business dataframe

In [354]:
df_Business.columns

Index(['Gender', 'Customer Type', 'Age', 'Type of Travel', 'Flight Distance',
       'Inflight wifi service', 'Departure/Arrival time convenient',
       'Ease of Online booking', 'Food and drink', 'Inflight entertainment',
       'Baggage handling', 'Cleanliness', 'Departure Delay in Minutes',
       'Arrival Delay in Minutes', 'Satisfaction'],
      dtype='object')

#### Renaming df_Business Columns to lower case and replacing blank spaces with '_'

In [356]:
df_Business.rename(columns = lambda x: x.lower().replace(' ','_'), inplace= True)
df_Business.columns

Index(['gender', 'customer_type', 'age', 'type_of_travel', 'flight_distance',
       'inflight_wifi_service', 'departure/arrival_time_convenient',
       'ease_of_online_booking', 'food_and_drink', 'inflight_entertainment',
       'baggage_handling', 'cleanliness', 'departure_delay_in_minutes',
       'arrival_delay_in_minutes', 'satisfaction'],
      dtype='object')

#### df_Business dataframe's information

In [358]:
df_Business.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 50000 entries, 0 to 49999
Data columns (total 15 columns):
 #   Column                             Non-Null Count  Dtype  
---  ------                             --------------  -----  
 0   gender                             48239 non-null  object 
 1   customer_type                      48239 non-null  object 
 2   age                                47373 non-null  float64
 3   type_of_travel                     48239 non-null  object 
 4   flight_distance                    46604 non-null  float64
 5   inflight_wifi_service              48239 non-null  float64
 6   departure/arrival_time_convenient  48239 non-null  float64
 7   ease_of_online_booking             42705 non-null  object 
 8   food_and_drink                     42705 non-null  object 
 9   inflight_entertainment             43649 non-null  float64
 10  baggage_handling                   43649 non-null  object 
 11  cleanliness                        43649 non-null  flo

## Modify improperly formatted data.
- Fixing the data format for categorical fields:
   - gender                             
   - customer_type
   - type_of_travel
   - satisfaction                       

In [360]:
# Printing unique values for list of columns
# for each column
for col in ['gender', 'customer_type','type_of_travel','satisfaction']:
    print(col)

    # get a list of unique values
    unique = df_Business[col].unique()
    print('#.of unique values: ', len(unique), '\nList of unique values: ',unique, '\n*************************************\n')


gender
#.of unique values:  6 
List of unique values:  ['Male' 'Female' 'F' 'M' nan 'Fe male'] 
*************************************

customer_type
#.of unique values:  5 
List of unique values:  ['Loyal Customer' 'Disloyal Customer' nan 'disloyal' 'Loyal'] 
*************************************

type_of_travel
#.of unique values:  4 
List of unique values:  ['Personal Travel' 'Business travel' 'Business' nan] 
*************************************

satisfaction
#.of unique values:  7 
List of unique values:  ['Dissatisfied' 'neutral' 'satisfied' 'dissatisfied' nan '0' '1'] 
*************************************



### Replacing values

In [362]:
df_Business.isnull().sum()

gender                               1761
customer_type                        1761
age                                  2627
type_of_travel                       1761
flight_distance                      3396
inflight_wifi_service                1761
departure/arrival_time_convenient    1761
ease_of_online_booking               7295
food_and_drink                       7295
inflight_entertainment               6351
baggage_handling                     6351
cleanliness                          6351
departure_delay_in_minutes           6351
arrival_delay_in_minutes             7902
satisfaction                         2160
dtype: int64

In [364]:
df_Business['gender'] = df_Business['gender'].replace(['M', 'F', 'Fe male'],['Male', 'Female', 'Female'])
df_Business['customer_type'] = df_Business['customer_type'].replace(['Loyal','disloyal'],['Loyal Customer','Disloyal Customer'])
df_Business['type_of_travel'] = df_Business['type_of_travel'].replace(['Business','Business travel'],['Business Travel','Business Travel'])
df_Business['satisfaction'] = df_Business['satisfaction'].replace(['dissatisfied', 'neutral', 'satisfied'], ['Dissatisfied', 'Neutral', 'Satisfied'])

In [366]:
df_Business.isnull().sum()

gender                               1761
customer_type                        1761
age                                  2627
type_of_travel                       1761
flight_distance                      3396
inflight_wifi_service                1761
departure/arrival_time_convenient    1761
ease_of_online_booking               7295
food_and_drink                       7295
inflight_entertainment               6351
baggage_handling                     6351
cleanliness                          6351
departure_delay_in_minutes           6351
arrival_delay_in_minutes             7902
satisfaction                         2160
dtype: int64

#### Printing unique values for list of columns

In [368]:
for col in ['gender', 'customer_type','type_of_travel','satisfaction']:
    print(col)

    # get a list of unique values
    unique = df_Business[col].unique()
    print('#.of unique values: ', len(unique), '\nList of unique values: ',unique, '\n*************************************\n')


gender
#.of unique values:  3 
List of unique values:  ['Male' 'Female' nan] 
*************************************

customer_type
#.of unique values:  3 
List of unique values:  ['Loyal Customer' 'Disloyal Customer' nan] 
*************************************

type_of_travel
#.of unique values:  3 
List of unique values:  ['Personal Travel' 'Business Travel' nan] 
*************************************

satisfaction
#.of unique values:  6 
List of unique values:  ['Dissatisfied' 'Neutral' 'Satisfied' nan '0' '1'] 
*************************************



## Droping fully duplicate observations.

In [370]:
# Print the number of duplicate observations.
df_Business.duplicated().sum()

3065

In [372]:
#Drop Duplicate Observations.
df_Business.drop_duplicates(inplace=True)

In [374]:
print("Re-checking the number of duplicate observations:")
df_Business.duplicated().sum()

Re-checking the number of duplicate observations:


0

## Handling missing values in all columns

#### Number of missing values in each column after dropping fully duplicate observations

In [376]:
df_Business.isnull().sum()

gender                                  2
customer_type                           2
age                                   863
type_of_travel                          2
flight_distance                       832
inflight_wifi_service                   2
departure/arrival_time_convenient       2
ease_of_online_booking               4747
food_and_drink                       4747
inflight_entertainment               4442
baggage_handling                     4442
cleanliness                          4442
departure_delay_in_minutes           4442
arrival_delay_in_minutes             5282
satisfaction                         2131
dtype: int64

#### Droping the least number of missing values data

In [378]:
columns_to_drop_na = ['gender', 'customer_type', 'type_of_travel', 
                       'inflight_wifi_service', 'departure/arrival_time_convenient']

df_Business.dropna(subset=columns_to_drop_na, inplace=True)

In [380]:
df_Business.isnull().sum()

gender                                  0
customer_type                           0
age                                   861
type_of_travel                          0
flight_distance                       830
inflight_wifi_service                   0
departure/arrival_time_convenient       0
ease_of_online_booking               4745
food_and_drink                       4745
inflight_entertainment               4440
baggage_handling                     4440
cleanliness                          4440
departure_delay_in_minutes           4440
arrival_delay_in_minutes             5280
satisfaction                         2131
dtype: int64

#### Handling the missing values with Mode

In [382]:
cols = df_Business.columns
for col in cols:
    print(col)
    value_counts = df_Business[col].value_counts()
    print(value_counts, '\n\n=======================================================')

gender
gender
Female    24058
Male      22875
Name: count, dtype: int64 

customer_type
customer_type
Loyal Customer       39656
Disloyal Customer     7277
Name: count, dtype: int64 

age
age
39.0     1913
40.0     1659
41.0     1511
42.0     1501
44.0     1491
         ... 
73.0       22
78.0       15
85.0       14
180.0       1
185.0       1
Name: count, Length: 74, dtype: int64 

type_of_travel
type_of_travel
Business Travel    44908
Personal Travel     2025
Name: count, dtype: int64 

flight_distance
flight_distance
337.0        222
2475.0       183
2586.0       147
404.0        133
214.0        132
            ... 
1221.0         1
1215.0         1
1203.0         1
1198.0         1
9000000.0      1
Name: count, Length: 3574, dtype: int64 

inflight_wifi_service
inflight_wifi_service
2.0    10507
3.0    10298
4.0     8882
1.0     8544
5.0     6846
0.0     1856
Name: count, dtype: int64 

departure/arrival_time_convenient
departure/arrival_time_convenient
4.0    9697
3.0    8953
5.0

In [384]:
columns_to_impute_mode = ['ease_of_online_booking', 'food_and_drink', 'inflight_entertainment', 'baggage_handling', 'cleanliness', 'satisfaction']

for col in columns_to_impute_mode:
    mode_value = df_Business[col].mode()[0]
    df_Business.fillna({col: mode_value}, inplace=True)

In [386]:
df_Business.isnull().sum()

gender                                  0
customer_type                           0
age                                   861
type_of_travel                          0
flight_distance                       830
inflight_wifi_service                   0
departure/arrival_time_convenient       0
ease_of_online_booking                  0
food_and_drink                          0
inflight_entertainment                  0
baggage_handling                        0
cleanliness                             0
departure_delay_in_minutes           4440
arrival_delay_in_minutes             5280
satisfaction                            0
dtype: int64

#### Handling the missing values with Mean

In [392]:
df_Business['age'] = pd.to_numeric(df_Business['age'], errors='coerce')
df_Business['flight_distance'] = pd.to_numeric(df_Business['flight_distance'], errors='coerce')
df_Business['departure_delay_in_minutes'] = pd.to_numeric(df_Business['departure_delay_in_minutes'], errors='coerce')
df_Business['arrival_delay_in_minutes'] = pd.to_numeric(df_Business['arrival_delay_in_minutes'], errors='coerce')

# Setting conditions for outlier removal
# Remove rows where age is not between 0 and 100
df_Business = df_Business[(df_Business['age'] > 0) & (df_Business['age'] < 100)]

# Assuming a the longest flight path ever acheived as upper limit for flight distance;
df_Business = df_Business[(df_Business['flight_distance'] <= 9500) & (df_Business['flight_distance'] >= 1)]

# Remove rows where departure delay and arrival delay are negative
df_Business = df_Business[(df_Business['departure_delay_in_minutes'] >= 0) & (df_Business['arrival_delay_in_minutes'] >= 0)]
columns_to_impute_mean = ['age', 'flight_distance', 'departure_delay_in_minutes', 'arrival_delay_in_minutes']

for col in columns_to_impute_mean:
    mean_value = df_Business[col].mean()
    df_Business.fillna({col: mean_value}, inplace=True)

In [394]:
df_Business.isnull().sum()

gender                               0
customer_type                        0
age                                  0
type_of_travel                       0
flight_distance                      0
inflight_wifi_service                0
departure/arrival_time_convenient    0
ease_of_online_booking               0
food_and_drink                       0
inflight_entertainment               0
baggage_handling                     0
cleanliness                          0
departure_delay_in_minutes           0
arrival_delay_in_minutes             0
satisfaction                         0
dtype: int64

In [396]:
cols = df_Business.columns
for col in cols:
    print(col)
    value_counts = df_Business[col].value_counts()
    print(value_counts, '\n\n=======================================================')

gender
gender
Female    20351
Male      19093
Name: count, dtype: int64 

customer_type
customer_type
Loyal Customer       34540
Disloyal Customer     4904
Name: count, dtype: int64 

age
age
39.0    1712
40.0    1525
41.0    1402
44.0    1381
42.0    1376
        ... 
73.0      19
76.0      18
79.0      16
85.0      13
78.0      13
Name: count, Length: 64, dtype: int64 

type_of_travel
type_of_travel
Business Travel    37922
Personal Travel     1522
Name: count, dtype: int64 

flight_distance
flight_distance
337.0     193
2475.0    160
2586.0    132
214.0     117
404.0     117
         ... 
840.0       1
3564.0      1
809.0       1
801.0       1
32.0        1
Name: count, Length: 3540, dtype: int64 

inflight_wifi_service
inflight_wifi_service
2.0    8766
3.0    8647
4.0    7539
1.0    6927
5.0    5993
0.0    1572
Name: count, dtype: int64 

departure/arrival_time_convenient
departure/arrival_time_convenient
4.0    8146
3.0    7581
5.0    7548
2.0    7495
1.0    7069
0.0    1605
Name:

In [402]:
columns_to_int_Datatype = ['inflight_wifi_service', 
    'departure/arrival_time_convenient', 
    'ease_of_online_booking', 
    'food_and_drink', 
    'inflight_entertainment', 
    'cleanliness','age','baggage_handling']

for col in columns_to_int_Datatype:
    df_Business[col] = df_Business[col].astype(int)

In [404]:
# Columns with rating scales where 0 is rated can be replaced with lowest rating value i.e., 1
columns_to_correct = [
    'inflight_wifi_service', 
    'departure/arrival_time_convenient', 
    'ease_of_online_booking', 
    'food_and_drink', 
    'inflight_entertainment', 
    'cleanliness'
]
for col in columns_to_correct:
    df_Business[col] = df_Business[col].replace(0, 1)


In [408]:
# Columns with rating scales where 10 is rated can be replaced with highest rating value i.e., 5
columns_to_correct = ['inflight_entertainment']
for col in columns_to_correct:
    df_Business[col] = df_Business[col].replace(10, 5)

In [412]:
df_Business = df_Business[df_Business['satisfaction'].isin(['Satisfied', 'Dissatisfied', 'Neutral'])]

### Fixing data types

In [416]:
df_Business = df_Business.convert_dtypes()

In [418]:
cols = df_Business.columns
for col in cols:
    print(col)
    value_counts = df_Business[col].value_counts()
    print(value_counts, '\n\n=======================================================')

gender
gender
Female    20349
Male      19092
Name: count, dtype: Int64 

customer_type
customer_type
Loyal Customer       34537
Disloyal Customer     4904
Name: count, dtype: Int64 

age
age
39    1712
40    1525
41    1402
44    1381
42    1376
      ... 
73      18
76      18
79      16
85      13
78      13
Name: count, Length: 64, dtype: Int64 

type_of_travel
type_of_travel
Business Travel    37919
Personal Travel     1522
Name: count, dtype: Int64 

flight_distance
flight_distance
337     193
2475    160
2586    132
214     117
404     117
       ... 
889       1
886       1
3763      1
870       1
32        1
Name: count, Length: 3540, dtype: Int64 

inflight_wifi_service
inflight_wifi_service
2    8766
3    8647
1    8498
4    7537
5    5993
Name: count, dtype: Int64 

departure/arrival_time_convenient
departure/arrival_time_convenient
1    8674
4    8144
3    7581
5    7547
2    7495
Name: count, dtype: Int64 

ease_of_online_booking
ease_of_online_booking
1    8226
3    8092

In [421]:
df_Business.info()

<class 'pandas.core.frame.DataFrame'>
Index: 39441 entries, 0 to 49998
Data columns (total 15 columns):
 #   Column                             Non-Null Count  Dtype 
---  ------                             --------------  ----- 
 0   gender                             39441 non-null  string
 1   customer_type                      39441 non-null  string
 2   age                                39441 non-null  Int32 
 3   type_of_travel                     39441 non-null  string
 4   flight_distance                    39441 non-null  Int64 
 5   inflight_wifi_service              39441 non-null  Int32 
 6   departure/arrival_time_convenient  39441 non-null  Int32 
 7   ease_of_online_booking             39441 non-null  Int32 
 8   food_and_drink                     39441 non-null  Int32 
 9   inflight_entertainment             39441 non-null  Int32 
 10  baggage_handling                   39441 non-null  Int32 
 11  cleanliness                        39441 non-null  Int32 
 12  departure

In [425]:
# Reset the index of the DataFrame
df_Business.reset_index(drop=True, inplace = True)
df_Business

gender   customer_type  age   type_of_travel  flight_distance  \
0        Male  Loyal Customer   69  Personal Travel               32   
1      Female  Loyal Customer   39  Business Travel               56   
2      Female  Loyal Customer   33  Business Travel               56   
3        Male  Loyal Customer   55  Business Travel               67   
4      Female  Loyal Customer   56  Business Travel               67   
...       ...             ...  ...              ...              ...   
39436  Female  Loyal Customer   37  Business Travel             4983   
39437    Male  Loyal Customer   45  Business Travel             3305   
39438  Female  Loyal Customer   35  Business Travel              337   
39439    Male  Loyal Customer   47  Business Travel             3496   
39440    Male  Loyal Customer   42  Business Travel             3605   

       inflight_wifi_service  departure/arrival_time_convenient  \
0                          3                                  2   
1                          2                                  5   
2                          4                                  5   
3                          5                                  5   
4                          1                                  4   
...                      ...                                ...   
39436                      2                                  3   
39437                      2                                  2   
39438                      5                                  5   
39439                      2                                  2   
39440                      4                                  4   

       ease_of_online_booking  food_and_drink  inflight_entertainment  \
0                           3               5                       1   
1                           5               5                       2   
2                           4               1                       4   
3                           5               1                       5   
4                           4               1                       3   
...                       ...             ...                     ...   
39436                       2               4                       3   
39437                       2               3                       4   
39438                       5               2                       4   
39439                       2               2                       5   
39440                       4               3                       4   

       baggage_handling  cleanliness  departure_delay_in_minutes  \
0                     1            3                          14   
1                     5            5                           0   
2                     4            2                           0   
3                     5            2                          10   
4                     1            3                           0   
...                 ...          ...                         ...   
39436                 1            4                           0   
39437                 4            3                           8   
39438                 4            5                          54   
39439                 5            4                           0   
39440                 4            3                           0   

       arrival_delay_in_minutes  satisfaction  
0                             0  Dissatisfied  
1                             0     Satisfied  
2                             0     Satisfied  
3                            12     Satisfied  
4                             0  Dissatisfied  
...                         ...           ...  
39436                         2     Satisfied  
39437                         0     Satisfied  
39438                        40     Satisfied  
39439                         2     Satisfied  
39440                         0     Satisfied  

[39441 rows x 15 columns]

#### Unique values for each column after cleansing

In [428]:
# for each column
cols = df_Business.columns
for col in cols:
    print(col)

    # get a list of unique values
    unique = df_Business[col].unique()
    print(unique, '\n*************************************\n\n')

gender
<StringArray>
['Male', 'Female']
Length: 2, dtype: string 
*************************************


customer_type
<StringArray>
['Loyal Customer', 'Disloyal Customer']
Length: 2, dtype: string 
*************************************


age
<IntegerArray>
[69, 39, 33, 55, 56, 47, 54, 57, 25, 51, 60, 27, 20, 19, 63, 50, 38, 46, 85,
 59, 41, 26, 42, 22, 52, 17, 48, 16, 53, 32, 35, 45, 44, 49, 65, 67, 66, 64,
 21, 37, 36, 15, 62, 40, 43, 68, 70, 58, 34, 11, 74, 13, 80, 75, 71, 12, 61,
 73, 72, 79, 76, 14, 77, 78]
Length: 64, dtype: Int32 
*************************************


type_of_travel
<StringArray>
['Personal Travel', 'Business Travel']
Length: 2, dtype: string 
*************************************


flight_distance
<IntegerArray>
[  32,   56,   67,   73,   74,   76,   77,   78,   80,   82,
 ...
 3996, 3997, 3998, 3999, 4000, 4243, 4502, 4817, 4963, 4983]
Length: 3540, dtype: Int64 
*************************************


inflight_wifi_service
<IntegerArray>
[3, 2, 4, 5, 1]
Le

In [430]:
df_Business

gender   customer_type  age   type_of_travel  flight_distance  \
0        Male  Loyal Customer   69  Personal Travel               32   
1      Female  Loyal Customer   39  Business Travel               56   
2      Female  Loyal Customer   33  Business Travel               56   
3        Male  Loyal Customer   55  Business Travel               67   
4      Female  Loyal Customer   56  Business Travel               67   
...       ...             ...  ...              ...              ...   
39436  Female  Loyal Customer   37  Business Travel             4983   
39437    Male  Loyal Customer   45  Business Travel             3305   
39438  Female  Loyal Customer   35  Business Travel              337   
39439    Male  Loyal Customer   47  Business Travel             3496   
39440    Male  Loyal Customer   42  Business Travel             3605   

       inflight_wifi_service  departure/arrival_time_convenient  \
0                          3                                  2   
1                          2                                  5   
2                          4                                  5   
3                          5                                  5   
4                          1                                  4   
...                      ...                                ...   
39436                      2                                  3   
39437                      2                                  2   
39438                      5                                  5   
39439                      2                                  2   
39440                      4                                  4   

       ease_of_online_booking  food_and_drink  inflight_entertainment  \
0                           3               5                       1   
1                           5               5                       2   
2                           4               1                       4   
3                           5               1                       5   
4                           4               1                       3   
...                       ...             ...                     ...   
39436                       2               4                       3   
39437                       2               3                       4   
39438                       5               2                       4   
39439                       2               2                       5   
39440                       4               3                       4   

       baggage_handling  cleanliness  departure_delay_in_minutes  \
0                     1            3                          14   
1                     5            5                           0   
2                     4            2                           0   
3                     5            2                          10   
4                     1            3                           0   
...                 ...          ...                         ...   
39436                 1            4                           0   
39437                 4            3                           8   
39438                 4            5                          54   
39439                 5            4                           0   
39440                 4            3                           0   

       arrival_delay_in_minutes  satisfaction  
0                             0  Dissatisfied  
1                             0     Satisfied  
2                             0     Satisfied  
3                            12     Satisfied  
4                             0  Dissatisfied  
...                         ...           ...  
39436                         2     Satisfied  
39437                         0     Satisfied  
39438                        40     Satisfied  
39439                         2     Satisfied  
39440                         0     Satisfied  

[39441 rows x 15 columns]

# Export the cleaned dataset to CSV file.

In [433]:
#Exporting data into csv file
df_Business.to_csv("Project_2_Group6.csv", index= False)

# Comparing the Data Characteristics Before and After Data Cleansing.

### Data frame(df_Business) after data cleansing. 
##### Observed changes:

Total data points were reduced from 104910 to 50000 as we filterd the data based on "Class" = "Business" and post cleansing it is further reduced to 39441.
The column values of the dataset are handled as per the Data dictionary standards.
Initially there were 16 columns but the class columns was dropped as per the expectations.
There were 3065 totally duplicate data points in the dataset which were cleaned.
Merged similar categories into one. for gender, customer_type, type_of_travel and satisfaction.
Columns such as 'gender', 'customer_type', 'type_of_travel', 'inflight_wifi_service', 'departure/arrival_time_convenient' had the least number of missing/null values, hence were dropped.
Columns such as 'ease_of_online_booking', 'food_and_drink', 'inflight_entertainment', 'baggage_handling', 'cleanliness', 'satisfaction': missing/null values were replaced by mode.
Columns such as 'age','flight_distance','departure_delay_in_minutes' and 'arrival_delay_in_minutes': missing/null values were replaced by mean.

-Column Characteristics Before and After Data Cleansing:

Gender:
Before: Object, 1761 missing values
After: String, no missing values
Changes: Missing values removed; datatype updated.

Customer Type:
Before: Object, 1761 missing values
After: String, no missing values
Changes: Missing values removed; datatype updated.

Age:
Before: Float64, 2627 missing values
After: Int32, no missing values
Changes: Outliers removed; mean imputed; datatype updated to integer.

Type of Travel:
Before: Object, 1761 missing values
After: String, no missing values
Changes: Missing values removed; datatype updated.

Flight Distance:
Before: Float64, 3396 missing values
After: Int64, no missing values
Changes: Outliers removed; mean imputed; datatype updated to integer.

Inflight Wi-Fi Service:
Before: Float64, 1761 missing values
After: Int32, no missing values
Changes: 0 values converted to 1; mode imputed; datatype updated.

Departure/Arrival Time Convenient:
Before: Float64, 1761 missing values
After: Int32, no missing values
Changes: 0 values converted to 1; mode imputed; datatype updated.

Ease of Online Booking:
Before: Object, 7295 missing values
After: Int32, no missing values
Changes: 0 values converted to 1; mode imputed; datatype updated.

Food and Drink:
Before: Object, 7295 missing values
After: Int32, no missing values
Changes: 0 values converted to 1; mode imputed; datatype updated.

Inflight Entertainment:
Before: Float64, 6351 missing values
After: Int32, no missing values
Changes: 0 values converted to 1; outliers removed; mode imputed; datatype updated.

Baggage Handling:
Before: Object, 6351 missing values
After: Int32, no missing values
Changes: Mode imputed; datatype updated.

Cleanliness:
Before: Float64, 6351 missing values
After: Int32, no missing values
Changes: 0 values converted to 1; mode imputed; datatype updated.

Departure Delay in Minutes:
Before: Float64, 6351 missing values
After: Int64, no missing values
Changes: Negative values removed; mean imputed; datatype updated.

Arrival Delay in Minutes:
Before: Object, 7902 missing values
After: Int64, no missing values
Changes: Negative values removed; mean imputed; datatype updated.

Satisfaction:
Before: Object, 2160 missing values
After: String, no missing values
Changes: Rows with invalid values (0, 1) removed; datatype updated.

Key Observations:
Missing Values: All columns were cleaned of missing values.
Outlier Removal: Logical outliers were removed from numeric columns.
Invalid Ratings: Rating columns with 0 values were adjusted to start from 1.
Datatype Standardization: Consistent datatypes were applied to facilitate analysis.

# * Before Cleansing *

### Uniqe Values for each Column

### Number of Missing Valuesfor each Column

# *** After Cleansing ***

### Columns names and datatype

### Uniqe Values for each Column

### Number of Missing Valuesfor each Column